torch/distributed/rpc/api.py

```diff
@@ -105,29 +105,28 @@ def __init__(self):
 
 
 def _on_leader_follower_report_shutdown_intent(sequence_id, worker_name):
-    assert (
-        worker_name in _ALL_WORKER_NAMES
-    ), "{worker_name} is not expected by leader.".format(worker_name=worker_name)
-    intent_worker_names = _wait_all_workers_sequence_id_to_states[
-        sequence_id
-    ].intent_worker_names
-    assert (
-        worker_name not in intent_worker_names
-    ), "{worker_name} reported intent sequence id {sequence_id} twice. ".format(
-        worker_name=worker_name, sequence_id=sequence_id
-    )
-    intent_worker_names.add(worker_name)
-    if _ALL_WORKER_NAMES == intent_worker_names:
-        _set_proceed_shutdown_signal(sequence_id)
+    with _wait_all_workers_dict_lock:
```

Comment:
Curious, wouldn't GIL already serve as a lock here?

Reply:
The interpreter actually releases GIL occasionally. Although we call the UDF while holding GIL, while executing Python code the interpreter might decide to release GIL and let another python thread execute something. Some context: https://stackoverflow.com/a/23855229